implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+use crate::secure_channel::handshake::error::XXError;
+use crate::secure_channel::handshake::handshake_state_machine::{HandshakeKeys, Status};
+use crate::secure_channel::Role;
+use crate::XXVault;
+use arrayref::array_ref;
+use ockam_core::compat::sync::Arc;
+use ockam_core::compat::{string::ToString, vec::Vec};
+use ockam_core::Result;
+use ockam_vault::constants::{AES256_SECRET_LENGTH_USIZE, CURVE25519_PUBLIC_LENGTH_USIZE};
+use ockam_vault::SecretType::X25519;
+use ockam_vault::{KeyId, PublicKey, Secret, SecretAttributes};
+use sha2::{Digest, Sha256};
+use Status::*;
+
+/// The number of bytes in a SHA256 digest
+pub const SHA256_SIZE_U32: u32 = 32;
+/// The number of bytes in a SHA256 digest
+pub const SHA256_SIZE_USIZE: usize = 32;
+/// The number of bytes in AES-GCM tag
+pub const AES_GCM_TAGSIZE_USIZE: usize = 16;
+
+/// Implementation of a Handshake for the noise protocol
+/// The first members are used in the implementation of some of the protocol steps, for example to
+/// encrypt messages
+/// The variables used in the protocol itself: s, e, rs, re,... are handled in `HandshakeState`
+pub(super) struct Handshake {
+    vault: Arc<dyn XXVault>,
+    pub(super) state: HandshakeState,
+}
+
+/// Top-level functions used in the initiator and responder state machines
+/// Each function makes mutable copy of the state to modify it in order to make the code more compact
+/// and avoid self.state.xxx = ...
+impl Handshake {
+    /// Initialize the handshake variables
+    pub(super) async fn initialize(&mut self) -> Result<()> {
+        let mut state = self.state.clone();
+        state.h = *Self::protocol_name();
+        state.k = self
+            .import_k_secret(vec![0u8; AES256_SECRET_LENGTH_USIZE])
+            .await?;
+        state.ck = self
+            .import_ck_secret(Self::protocol_name().to_vec())
+            .await?;
+
+        state.h = HandshakeState::sha256(&state.h);
+        self.state = state;
+        Ok(())
+    }
+
+    /// Encode the first message, sent from the initiator to the responder
+    pub(super) async fn encode_message1(&mut self) -> Result<Vec<u8>> {
+        let mut state = self.state.clone();
+        // output e.pubKey
+        let e_pub_key = self.get_public_key(&state.e).await?;
+        state.mix_hash(e_pub_key.data());
+        let mut message = e_pub_key.data().to_vec();
```

Comment:
Can we at least have a check for endianness and throw an error if it's little-endian?

Reply:
What is the fireproof way to test that?

Reply:
https://doc.rust-lang.org/reference/conditional-compilation.html#target_endian

Reply:
Maybe I misunderstood your point. I thought you wanted to test if the sent data was sent by a big or little endian system. You actually want to make sure that the local system is little endian and say that we don't support big endian systems?

Reply:
I meant that there is extra work we need to do to make this work correctly on big-endian. Since we don't want to spend time on it right now, a good enough solution would be to not compile on big-endian (which is better than compile but work incorrectly).